content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown

```diff
@@ -0,0 +1,173 @@
+
+Why did we decide to move away from these functions in favour of `across()`?
+
+1.  `across()` makes it possible to compute useful summaries that were 
+    previously impossible. For example, it's now trivial to summarise
+    numeric vectors with one function, factors with another, and still 
+    compute the number of rows in each group:
+
+    ```{r, eval = FALSE}
+    df %>%
+      group_by(g1, g2) %>% 
+      summarise(
+        across(is.numeric, mean), 
+        across(is.factor, nlevels),
+        n = n(), 
+      )
+    ```
+
+2.  `across()` reduces the number of functions that dplyr needs to provide. 
+    This makes dplyr easier for you to use (because there are fewer functions 
+    to remember) and easier for us to develop (since we only need to implement 
+    one function for each new verb, not four).
+
+3.  `across()` unifies `_if` and `_at` semantics, allowing combinations that 
+    used to be impossible. For example, you can now transform all numeric 
+    columns whose name begins with "x": `across(is.numeric & starts_with("x"))`.
+
+4.  `across()` doesn't need `vars()`. The `_at()` functions are the only place 
+    in dplyr where you have to use `vars()`, which makes them unusual, 
+    and hence harder to learn and remember.
+
+Why did it take it long to discover `across()`? Surprisingly, the key idea that makes `across()` works came out of our low-level work on the [vctrs](http://vctrs.r-lib.org/) package, where we learnt that you can have a column of a data frame that is itself a data frame. It's a bummer that we had a few false starts before we discovered `across()`, but even with hindsight, I don't see how we could've skipped the intermediate steps.
+
+## How do you convert existing code?
+
+Fortunately, if you have code that uses the `_if`, `_at`, or `_all()` functions, it's generally straightforward to update:
+
+*   Strip the `_if()`, `_at()` and `_all()` suffix off the function.
+
+*   Call `across()`. The first argument will be:
+
+    1. For `_if()`, the old second argument.
+    1. For `_at()`, the old second argument, with the call to `vars()` removed.
+    1. For `_all()`, `everything()`.
+
+    The subsequent arguments can be copied as is.
+
+Here are a few examples of this process:
+
+```{r, eval = FALSE}
+df %>% mutate_if(is.numeric, mean, na.rm = TRUE)
+# ->
+df %>% mutate(across(is.numeric, mean, na.rm = TRUE))
+
+df %>% mutate_at(vars(c(x, starts_with("y"))), mean, na.rm = TRUE)
+# ->
+df %>% mutate(across(c(x, starts_with("y")), mean, na.rm = TRUE))
+
+df %>% mutate_all(mean, na.rm = TRUE)
+# ->
+df %>% mutate(across(everything(), mean, na.rm = TRUE))
+```
+
+If you've used multiple `_if`/`_at`/`_all` functions in a row, you should also consider if it's now possible to collapse into a single call, using the new features of `across()`.
+
+Again, you don't need to worry about these functions going away in the short-term, but it's good practice to update your code to use our latest recommendations.
```

Comment:
> but it's good practice to update your code to use our latest recommendations.

Is there another way of saying this that alludes to the merits of `across()` rather than our recommendations?

Reply:
"use the latest features"?

That reminds me that I need to mention that `across()` is slower currently but we have a planning for improving performance in 1.1.0

Reply:
I like it.